server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -628,17 +629,23 @@ public String resolveDocumentType(String type) {
         return type;
     }
 
+    public DocumentMapper documentMapperWithAutoCreate(String type) {
+        return documentMapperWithAutoCreate(type, m -> {});
+    }
+
     /**
-     * Returns the document mapper created, including a mapping update if the
-     * type has been dynamically created.
+     * If the document mapper is already set, returns it.  If not, then creates
+     * one, and calls the updateConsumer with a mapping update before
+     * returning the new mapper.
      */
-    public DocumentMapperForType documentMapperWithAutoCreate(String type) {
+    public DocumentMapper documentMapperWithAutoCreate(String type, Consumer<Mapping> updateConsumer) {
```

Comment:
In terms of API, I'd rather like that we return an additional boolean that says whether the mapping was created or not? Maybe by replacing the return type by a `Tuple<DocumentMapper,Boolean>` if we don't want to introduce a new class. Then if the boolean is true, consumers can get the mapping update via `DocumentMapper#mapping`?